File: src/commands/bindchannel.rs
use crate::components::v2::FadeResponse;
use crate::error::{BotError, BotResult};
use crate::state::AppState;
use serenity::model::Permissions;
use serenity::prelude::*;
use std::sync::Arc;
use tokio::sync::RwLock;

pub async fn run(
    ctx: &Context,
    cmd: &crate::commands::context::CommandContext<'_>,
    state: Arc<RwLock<AppState>>,
    args: &[&str],
) -> BotResult<()> {
    cmd.defer(ctx).await?;

    let guild_id = cmd.guild_id().ok_or(BotError::Custom("This command can only be used in a server.".to_string()))?;

    // Check admin
    let has_admin = match cmd {
        crate::commands::context::CommandContext::Prefix(msg) => {
            if let Some(member) = &msg.member {
                member.permissions.unwrap_or(Permissions::empty()).contains(Permissions::ADMINISTRATOR)
            } else {
                false
            }
        },
        crate::commands::context::CommandContext::Slash(interaction) => {
            if let Some(member) = &interaction.member {
                member.permissions.unwrap_or(Permissions::empty()).contains(Permissions::ADMINISTRATOR)
            } else {
                false
            }
        }
    };

    if !has_admin {
        return Err(BotError::Custom("You need Administrator permissions to use this command.".to_string()));
    }

    let mut target_channel = cmd.channel_id();

    // Parse slash command options or prefix arguments for channel mention
    match cmd {
        crate::commands::context::CommandContext::Slash(interaction) => {
            if let Some(serenity::model::application::CommandDataOptionValue::Channel(id)) = interaction.data.options.iter().find(|opt| opt.name == "channel").map(|opt| opt.value.clone()) {
                target_channel = id;
            }
        },
        crate::commands::context::CommandContext::Prefix(_) => {
            if !args.is_empty() {
                let text = args[0].trim_matches(|c| c == '<' || c == '#' || c == '>');
                if let Ok(id) = text.parse::<u64>() {
                    target_channel = serenity::model::id::ChannelId::new(id);
                }
            }
        }
    }

    let is_whitelisted = {
        let state_guard = state.read().await;
        let mut db = state_guard.db.write().await;
        let list = db.whitelisted_channels.entry(guild_id.get()).or_default();
        
        let currently = list.contains(&target_channel.get());
        if currently {
            list.remove(&target_channel.get());
        } else {
            list.insert(target_channel.get());
        }
        
        db.save();
        !currently
    };

    use crate::components::emoji::E;
    let msg = if is_whitelisted {
        format!("{} <#{}> is now **whitelisted/bound**. I will ONLY listen to commands in bound channels.", E::OK, target_channel.get())
    } else {
        format!("{} <#{}> is **no longer whitelisted**.", E::OK, target_channel.get())
    };

    let card = FadeResponse::new().container(None, |c| c.text(msg));
    cmd.edit(ctx, &card).await?;

    Ok(())
}

File: ts/src/commands/bindchannel.ts
import { Declare, Command, type CommandContext, Options, createChannelOption } from 'seyfert';
import { ChannelType } from 'seyfert/lib/types/index';
import { appState } from '../state';
import { successCard, errorCard } from '../cards/common';
import { E } from '../components/emoji';

const options = {
  channel: createChannelOption({
    description: 'The channel to bind to (leave empty to unbind)',
    required: false,
    channel_types: [ChannelType.GuildText],
  }),
};

@Declare({
  name: 'bindchannel',
  description: 'Bind bot commands to a specific channel or clear the bind.',
})
@Options(options)
export default class BindChannelCommand extends Command {
  override async run(ctx: CommandContext<typeof options>) {
    if (!ctx.guildId) {
      await ctx.editOrReply(errorCard('This command can only be used in a server.').toMessage());
      return;
    }

    const state = appState();
    const channel = ctx.options.channel;

    if (!channel) {
      state.db.whitelistedChannels.delete(ctx.guildId);
      state.db.save();
      await ctx.editOrReply(successCard(`${E.NOTES} Unbound the bot. Commands can now be used anywhere.`).toMessage());
      return;
    }

    let whitelisted = state.db.whitelistedChannels.get(ctx.guildId);
    if (!whitelisted) {
      whitelisted = new Set();
      state.db.whitelistedChannels.set(ctx.guildId, whitelisted);
    }
    
    // In Rust it was typically a single bound channel or multiple. 
    // We will clear existing and set this one, to act like a strict bind.
    whitelisted.clear();
    whitelisted.add(channel.id);
    state.db.save();

    await ctx.editOrReply(successCard(`${E.NOTES} Bot bound to <#${channel.id}>.`).toMessage());
  }
}
